mtp_api/apps/prison/serializers.py

```diff
@@ -110,10 +128,27 @@ def get_combined_account_balance(self, prisoner_location: PrisonerLocation):
             msg = f'NOMIS balances for {prisoner_location.prisoner_number} is malformed: {e}'
             logger.exception(msg)
             raise ValidationError(msg)
-        except requests.RequestException:
-            msg = f'Cannot lookup NOMIS balances for {prisoner_location.prisoner_number}'
-            logger.exception(msg)
-            raise ValidationError(msg)
+        except requests.RequestException as e:
+            # TODO I think this message is rendered on page, make it nicer
+            msg = (
+                f'Cannot lookup NOMIS balances for {prisoner_location.prisoner_number} in '
+                f'{prisoner_location.prison.nomis_id}'
+            )
+            if (
+                getattr(e, 'response', None) is not None and
+                e.response.status_code == status.HTTP_400_BAD_REQUEST
+                and update_location_on_not_found
+            ):
+                logger.warning(msg)
+                new_location = PrisonerLocationSerializer(prisoner_location).fetch_and_update()
```

Comment:
I can't see any good reason not to. I'll add in the signal invocation

Reply:
i have a suspicion that when a prisoner location file is uploaded that there _might_ be some side effects other than just creating location models. it would be worth checking. it might be this signal is called or maybe it isn't even needed!

Reply:
I'll have a look into what the signal's doing

Reply:
So it looks like it's calling through to https://github.com/ministryofjustice/money-to-prisoners-api/blob/6e581e0451a6267ab95ae0b79a2181a2c4b26f73/mtp_api/apps/credit/managers.py#L48 which is traversing the credit table. I did a quick explain to see what the underlying query was doing and I came up with:
```
(ins)mtp_api=# EXPLAIN ANALYSE UPDATE credit_credit SET prison_id = pl.prison_id, prisoner_name = pl.prisoner_name, single_offender_id = pl.single_offender_id FROM credit_credit AS c LEFT OUTER JOIN pri
son_prisonerlocation AS pl ON c.prisoner_number = pl.prisoner_number AND c.prisoner_dob = pl.prisoner_dob AND pl.active IS True LEFT OUTER JOIN payment_payment AS p ON p.credit_id=c.id WHERE c.owner_id
IS NULL AND c.resolution = 'pending' AND c.reconciled is False AND credit_credit.id = c.id AND NOT (pl.prison_id IS NULL AND p.uuid IS NOT NULL);
                                                                   QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------
 Update on credit_credit  (cost=38.38..60.60 rows=23 width=283) (actual time=35.376..35.377 rows=0 loops=1)
   ->  Hash Left Join  (cost=38.38..60.60 rows=23 width=283) (actual time=13.417..13.974 rows=40 loops=1)
         Hash Cond: (((c.prisoner_number)::text = (pl.prisoner_number)::text) AND (c.prisoner_dob = pl.prisoner_dob))
         Filter: ((pl.prison_id IS NOT NULL) OR (p.uuid IS NULL))
         ->  Hash Join  (cost=34.13..56.16 rows=23 width=239) (actual time=12.501..12.970 rows=40 loops=1)
               Hash Cond: (credit_credit.id = c.id)
               ->  Seq Scan on credit_credit  (cost=0.00..20.22 rows=422 width=200) (actual time=10.001..10.388 rows=422 loops=1)
               ->  Hash  (cost=33.84..33.84 rows=23 width=43) (actual time=2.333..2.333 rows=40 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 11kB
                     ->  Hash Right Join  (cost=21.56..33.84 rows=23 width=43) (actual time=1.374..2.280 rows=40 loops=1)
                           Hash Cond: (p.credit_id = c.id)
                           ->  Seq Scan on payment_payment p  (cost=0.00..11.59 rows=259 width=26) (actual time=0.342..1.144 rows=301 loops=1)
                           ->  Hash  (cost=21.27..21.27 rows=23 width=21) (actual time=1.005..1.005 rows=40 loops=1)
                                 Buckets: 1024  Batches: 1  Memory Usage: 11kB
                                 ->  Seq Scan on credit_credit c  (cost=0.00..21.27 rows=23 width=21) (actual time=0.165..0.947 rows=40 loops=1)
                                       Filter: ((owner_id IS NULL) AND (reconciled IS FALSE) AND ((resolution)::text = 'pending'::text))
                                       Rows Removed by Filter: 382
         ->  Hash  (cost=3.50..3.50 rows=50 width=51) (actual time=0.807..0.807 rows=50 loops=1)
               Buckets: 1024  Batches: 1  Memory Usage: 13kB
               ->  Seq Scan on prison_prisonerlocation pl  (cost=0.00..3.50 rows=50 width=51) (actual time=0.520..0.732 rows=50 loops=1)
                     Filter: (active IS TRUE)
 Planning Time: 3.011 ms
 Execution Time: 35.670 ms
(23 rows)
```
At the moment you're right it will probably take too long to run synchronously. That last sequence scan can definitely be replaced by an index lookup though. I'll add that now before we forget about it, and we can profile how long it takes on a live-sized database. If it still takes too long (my feeling is that it might) then we can avoid updating the database.

Reply:
Yeah, we definitely won't be able to run `update_prisons` synchronously. It takes >4 seconds on a database 1/10th of the size of production, even with the extra indices i've added. This update is running over all prisons, so there's definitely optimisations to be had there, but I'm not entirely sure it's worth making that change now.
```
(ins)>>> print(s.getvalue())
         247 function calls (244 primitive calls) in 117.387 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        3    0.001    0.000  112.508   37.503 /usr/lib/python3.8/code.py:263(raw_input)
        3  112.507   37.502  112.507   37.502 {built-in method builtins.input}
        3    0.000    0.000    4.879    1.626 /usr/lib/python3.8/code.py:242(push)
        3    0.000    0.000    4.879    1.626 /usr/lib/python3.8/code.py:38(runsource)
        3    0.000    0.000    4.877    1.626 /usr/lib/python3.8/code.py:77(runcode)
        3    0.000    0.000    4.877    1.626 {built-in method builtins.exec}
        1    0.000    0.000    4.877    4.877 /home/tobe/Src/MadeTech/MOJ/MoneyToPrisoners/money-to-prisoners-api/mtp_api/apps/credit/managers.py:48(update_prisons)
        2    4.838    2.419    4.838    2.419 {method 'execute' of 'psycopg2.extensions.cursor' objects}
        1    0.000    0.000    4.835    4.835 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:97(execute)
        1    0.000    0.000    4.835    4.835 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:67(execute)
        1    0.000    0.000    4.835    4.835 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:73(_execute_with_wrappers)
        1    0.000    0.000    4.835    4.835 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:79(_execute)
        1    0.000    0.000    0.042    0.042 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:253(cursor)
        1    0.000    0.000    0.041    0.041 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:231(_cursor)
      4/1    0.000    0.000    0.041    0.041 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:212(ensure_connection)
        1    0.000    0.000    0.041    0.041 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:179(connect)
        1    0.000    0.000    0.038    0.038 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:167(get_new_connection)
        1    0.000    0.000    0.037    0.037 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/psycopg2/__init__.py:82(connect)
        1    0.036    0.036    0.036    0.036 {built-in method psycopg2._psycopg._connect}
        1    0.000    0.000    0.004    0.004 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:197(init_connection_state)
        1    0.000    0.000    0.004    0.004 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:187(ensure_timezone)
        3    0.000    0.000    0.002    0.001 /usr/lib/python3.8/codeop.py:152(__call__)
        3    0.000    0.000    0.002    0.001 /usr/lib/python3.8/codeop.py:68(_maybe_compile)
        9    0.000    0.000    0.002    0.000 /usr/lib/python3.8/codeop.py:135(__call__)
        9    0.002    0.000    0.002    0.000 {built-in method builtins.compile}
        2    0.000    0.000    0.001    0.001 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/psycopg2/extensions.py:146(make_dsn)
        2    0.001    0.000    0.001    0.000 {built-in method psycopg2._psycopg.parse_dsn}
        2    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/psycopg2/extensions.py:171(<listcomp>)
       10    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/psycopg2/extensions.py:180(_param_escape)
       10    0.000    0.000    0.000    0.000 {method 'sub' of 're.Pattern' objects}
       10    0.000    0.000    0.000    0.000 /usr/lib/python3.8/re.py:323(_subx)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/re.py:313(_compile_repl)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/sre_parse.py:969(parse_template)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/encodings/__init__.py:70(search_function)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/models/manager.py:26(__init__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/models/manager.py:20(__new__)
        2    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/utils/functional.py:28(__get__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/__init__.py:32(__getattr__)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/logging/__init__.py:1412(debug)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/logging/__init__.py:1677(isEnabledFor)
        3    0.000    0.000    0.000    0.000 /usr/lib/python3.8/encodings/utf_8.py:15(decode)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:145(get_connection_params)
        4    0.000    0.000    0.000    0.000 /usr/lib/python3.8/sre_parse.py:233(__next)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/encodings/__init__.py:43(normalize_encoding)
        3    0.000    0.000    0.000    0.000 /usr/lib/python3.8/sre_parse.py:254(get)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/encodings/utf_8.py:33(getregentry)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:381(set_autocommit)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/models/manager.py:119(_set_creation_counter)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/sre_parse.py:224(__init__)
        2    0.000    0.000    0.000    0.000 {method 'cursor' of 'psycopg2.extensions.connection' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/utils.py:195(__getitem__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:220(_prepare_cursor)
        1    0.000    0.000    0.000    0.000 <console>:1(<module>)
        3    0.000    0.000    0.000    0.000 {built-in method _codecs.utf_8_decode}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.__import__}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:37(__exit__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/operations.py:218(last_executed_query)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/sre_parse.py:978(addgroup)
       10    0.000    0.000    0.000    0.000 {method 'search' of 're.Pattern' objects}
        1    0.000    0.000    0.000    0.000 {method 'get_parameter_status' of 'psycopg2.extensions.connection' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:135(timezone_name)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:206(create_cursor)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/codecs.py:94(__new__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/base.py:226(_set_autocommit)
        7    0.000    0.000    0.000    0.000 {method 'join' of 'str' objects}
        4    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
        2    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/psycopg2/extensions.py:164(<dictcomp>)
       12    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:543(wrap_database_errors)
        3    0.000    0.000    0.000    0.000 {method 'split' of 'str' objects}
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/logging/__init__.py:214(_acquireLock)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        3    0.000    0.000    0.000    0.000 /usr/lib/python3.8/code.py:184(resetbuffer)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:518(validate_thread_sharing)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:558(make_debug_cursor)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/conf/__init__.py:53(__getattr__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:23(__getattr__)
        3    0.000    0.000    0.000    0.000 {built-in method time.time}
        1    0.000    0.000    0.000    0.000 {method 'update' of 'dict' objects}
        4    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
        2    0.000    0.000    0.000    0.000 {built-in method __new__ of type object at 0x8ff5e0}
        3    0.000    0.000    0.000    0.000 {method 'strip' of 'str' objects}
        5    0.000    0.000    0.000    0.000 {method 'isalnum' of 'str' objects}
        1    0.000    0.000    0.000    0.000 {method 'acquire' of '_thread.RLock' objects}
        4    0.000    0.000    0.000    0.000 {method 'items' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:201(check_settings)
        3    0.000    0.000    0.000    0.000 {method 'get' of 'dict' objects}
        2    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/logging/__init__.py:223(_releaseLock)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'set_client_encoding' of 'psycopg2.extensions.connection' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/utils.py:57(__init__)
        4    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/utils.py:68(__exit__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:376(get_autocommit)
        1    0.000    0.000    0.000    0.000 /usr/lib/python3.8/logging/__init__.py:1663(getEffectiveLevel)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:147(queries_logged)
        2    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
        4    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/  2    4.838    2.419    4.838    2.419 site-packages/django/db/utils.py:65(__enter__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:17(__init__)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/dispatch/dispatcher.py:155(send)
        1    0.000    0.000    0.000    0.000 {method 'replace' of 'str' objects}
        1    0.000    0.000    0.000    0.000 {built-in method _thread.get_ident}
        1    0.000    0.000    0.000    0.000 {method 'decode' of 'bytes' objects}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.chr}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:428(validate_no_atomic_block)
        1    0.000    0.000    0.000    0.000 {method 'release' of '_thread.RLock' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/base/base.py:434(validate_no_broken_transaction)
        1    0.000    0.000    0.000    0.000 {method 'close' of 'psycopg2.extensions.cursor' objects}
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/postgresql/operations.py:102(set_time_zone_sql)
        1    0.000    0.000    0.000    0.000 /home/tobe/.virtualenvs/money-to-prisoners-api/lib/python3.8/site-packages/django/db/backends/utils.py:34(__enter__)
        1    0.000    0.000    0.000    0.000 {method 'append' of 'collections.deque' objects}
```